federation/pkg/federation-controller/job/jobcontroller.go

```diff
@@ -0,0 +1,537 @@
+		return false
+	}
+
+	if !fjc.jobController.HasSynced() {
+		glog.V(2).Infof("federation job list not synced")
+		return false
+	}
+	return true
+}
+
+func (fjc *JobController) deliverLocalJob(obj interface{}, duration time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %v: %v", obj, err)
+		return
+	}
+	_, exists, err := fjc.jobStore.GetByKey(key)
+	if err != nil {
+		glog.Errorf("Couldn't get federated job %v: %v", key, err)
+		return
+	}
+	if exists { // ignore jobs exists only in local k8s
+		fjc.deliverJobByKey(key, duration, false)
+	}
+}
+
+func (fjc *JobController) deliverFedJobObj(obj interface{}, delay time.Duration) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	fjc.deliverJobByKey(key, delay, false)
+}
+
+func (fjc *JobController) deliverJobByKey(key string, delay time.Duration, failed bool) {
+	if failed {
+		fjc.jobBackoff.Next(key, time.Now())
+		delay = delay + fjc.jobBackoff.Get(key)
+	} else {
+		fjc.jobBackoff.Reset(key)
+	}
+	fjc.jobDeliverer.DeliverAfter(key, nil, delay)
+}
+
+type reconciliationStatus string
+
+const (
+	statusAllOk       = reconciliationStatus("ALL_OK")
+	statusNeedRecheck = reconciliationStatus("RECHECK")
+	statusError       = reconciliationStatus("ERROR")
+	statusNotSynced   = reconciliationStatus("NOSYNC")
+)
+
+func (fjc *JobController) worker() {
+	for {
+		item, quit := fjc.jobWorkQueue.Get()
+		if quit {
+			return
+		}
+		key := item.(string)
+		status, err := fjc.reconcileJob(key)
+		fjc.jobWorkQueue.Done(item)
+		if err != nil {
+			glog.Errorf("Error syncing job controller: %v", err)
+			fjc.deliverJobByKey(key, 0, true)
+		} else {
+			switch status {
+			case statusAllOk:
+				break
+			case statusError:
+				fjc.deliverJobByKey(key, 0, true)
+			case statusNeedRecheck:
+				fjc.deliverJobByKey(key, jobReviewDelay, false)
+			case statusNotSynced:
+				fjc.deliverJobByKey(key, clusterAvailableDelay, false)
+			default:
+				glog.Errorf("Unhandled reconciliation status: %s", status)
+				fjc.deliverJobByKey(key, jobReviewDelay, false)
+			}
+		}
+	}
+}
+
+type scheduleResult struct {
+	Parallelism *int32
+	Completions *int32
+}
+
+func (fjc *JobController) schedule(fjob *batchv1.Job, clusters []*fedv1.Cluster) map[string]scheduleResult {
+	plnr := fjc.defaultPlanner
+	frsPref, err := parseFedJobReference(fjob)
+	if err != nil {
+		glog.Warningf("Invalid job specific preference, use default. rs: %v, err: %v", fjob, err)
+	}
+	if frsPref != nil { // create a new planner if user specified a preference
+		plnr = planner.NewPlanner(frsPref)
+	}
+
+	parallelism := int64(*fjob.Spec.Parallelism)
+	var clusterNames []string
+	for _, cluster := range clusters {
+		clusterNames = append(clusterNames, cluster.Name)
+	}
+	parallelismResult, _ := plnr.Plan(parallelism, clusterNames, nil, nil, fjob.Namespace+"/"+fjob.Name)
```

Comment:
As per my comments in the design doc, I think that the solution here is to adjust downward the spec.completions of the job in the cluster where pods are failing to schedule, and transfer that number of completions to the job in another cluster where pods are successfully able to schedule. That way the total number of completions across all clusters cannot exceed the desired number.

I'm not sure what happens if one increases the number of completions for a job that has already completed it's allocated number of completions (and hence already shut down it's pods). One might hope that it will respawn pods to perform the additional required completions.  That would be first prize.  If this is not the case, it might be necessary to spawn a new Job for the additional completions required.  This raises the question of potential name clashes with the original completed but not yet deleted deleted job, but hopefully we can figure that out (e.g. by deleting the completed job) if that becomes necessary.